fix(migrations): control flow migration fails for async pipe with unboxing of observable

AFreeNeat opened the pull request:
Update control flow syntax to use 'as' for proper async pipe handling, accounting for variable whitespace before 'let'.

Fixes #52756

jessicajaniuk commented on packages/core/schematics/ng-generate/control-flow-migration/ifs.ts:
```diff
@@ -84,7 +84,10 @@ function migrateNgIf(etm: ElementToMigrate, tmpl: string, offset: number): Resul
   // includes the mandatory semicolon before as
   const lbString = etm.hasLineBreaks ? '\n' : '';
-  const condition = etm.attr.value.replace(' as ', '; as ');
+  const condition = etm.attr.value
```
Can you add this to the other 3 cases also? Lines 109, 116, and 154 would also need this fix.

AnthonyFr75 replied:
Hi @jessicajaniuk, thank you for your review.

Regarding line 109 for bound if else blocks, I don’t believe my fix is applicable. In this case, I think we should use a template input variable with let- as follows:
[ngIf]="logStatus$ | async" let-logStatus

jessicajaniuk replied:
@AnthonyFr75 I think it would apply to 109, but yeah, not 116. 109 applies to regular usages of if/else: `*ngIf="condition; else tmpl"`. 116 is for the bound usages of it that would have `[ngIf]` and `[ngIfElse]`.

AnthonyFr75 replied:
Hi @jessicajaniuk,

I just noticed that it would not apply to standard 'if-else-then' structures, either.
( even if declaring let is possible ).
In the code snippet:
`<div *ngIf="logStatus$ | async; let logStatus; then loggedIn else loggedOut"> Not displayed on the DOM, so logStatus is not be available </div>
<ng-template #loggedIn> Hello </ng-template>
<ng-template #loggedOut> Bye bye </ng-template>`
The logStatus input variable cannot be accessed by the templates. Additionally, the text inside the <div> is not displayed in the DOM because its rendering has been delegated to the templates.

jessicajaniuk replied:
@AnthonyFr75 Thanks for checking and verifying that!

jessicajaniuk commented on packages/core/schematics/test/control_flow_migration_spec.ts:
```diff
@@ -3408,5 +3407,24 @@ describe('control flow migration', () => {
+
+        @Component({
+          imports: [NgIf],
+          template: \`<ng-container *ngIf="value$ | async; let value"> {{value}} </ng-container>\`
```
If you want to add tests for the other scenarios (If/else, If/then/else, and NgIfElse) that would be great, but non-blocking. I think this covers the logic well enough.

AnthonyFr75 replied:
I add the missing test cases.

jessicajaniuk reviewed (changes requested):
This looks great. Just need to add it to the other spots, too.

jessicajaniuk commented:
@AnthonyFr75 Also, lint is failing because your commit is missing a body. While you're amending the commit, you could switch `fix(core)` to `fix(migrations)`. Both are technically fine, but the latter is more correct.

AFreeNeat committed:
fix(migrations): control flow migration fails for async pipe with unboxing of observable (#52756)

Update control flow syntax to use 'as' for proper async pipe handling, accounting for variable whitespace before 'let'.

Fixes #52756

jessicajaniuk reviewed (approved):
LGTM! Thanks for this!

AndrewKushnir commented:
This PR was merged into the repository by commit e33f6e0f1a483cad908fa6d7376d62332797499c.
